### PR TITLE
TLS auth for Acra-Translator

### PR DESCRIPTION
**Pull request by Lagovas**
**Current state**
Our Acra-Translator works with HTTP and gRPC requests.
Both of them require Acra-Connector with SecureSession, but differently work with TLS.

* In gRPC mode TLS works without AcraConnector and accept only direct connections from client's app. And due to that Translator in this mode doesn't use trace from client's side and start trace from itself.
With SecureSession it require Acra-Connector and continue trace from Acra-Connector, but doesn't use ClientID from handshake in request processing.

* In HTTP mode it works in another way. It always require Acra-Connector and work only with SecureSession as transport. After handshake it uses ClientID from handshake and pass to request processing and use it for encryption/decryption if ZoneID wasn't pass by request parameters.

**New state**

**HTTP**
In this PR added TLS support for HTTP mode to use with or without Acra-Connector. If used `--acraconnector_transport_encryption_disable` parameter then Acra-Translator doesn't expect trace id from client's app after handshake and it allows to connect by any HTTP client with TLS.
Otherwise should be used Acra-Connector configured to use TLS which will send trace id from start of session.
And in this setup ClientID will be used from TLS certificates.

**gRPC**

In gRPC mode extended TLS mode. If used `--acratranslator_client_id_from_connection_enable` parameter then ClientID extracted from TLS certificates and overrides ClientID in gRPC requests
Plus added SecureSession support with extracting ClientID from handshake if used `--acratranslator_client_id_from_connection_enable` parameter. But anyway gRPC doesn't accept trace id from Acra-Connector (in gRPC mode apps can't send it, in SecureSession it sends Acra-Connector but not used in Acra-Translator because handling connections under control of gRPC server implementation)

Please, pay attention to new arguments and their naming:  `--acratranslator_client_id_from_connection_enable` and  `--acraconnector_transport_encryption_disable`. Are they okay or maybe you can suggest better names?

For now not pass integration tests which okay pass locally and missing unit-tests which should be updated after some refactoring due to developing. I will switch PR from draft after fixing tests problem and add unit-tests

### CODE REVIEW

File: cmd/acra-translator/acra-translator.go [hunk 1]
```diff
@@ -80,6 +83,12 @@ func main() {
 
 	prometheusAddress := flag.String("incoming_connection_prometheus_metrics_string", "", "URL which will be used to expose Prometheus metrics (use <URL>/metrics address to pull metrics)")
 
+	useTLS := flag.Bool("acratranslator_tls_transport_enable", false, "Use tls to encrypt transport between AcraServer and AcraConnector/client")
```
Comment (vixentael) [inline cmd/acra-translator/acra-translator.go:86]:
```suggestion
	useTLS := flag.Bool("acratranslator_tls_transport_enable", false, "Use TLS to encrypt transport between AcraServer and AcraConnector/client app")
```

File: cmd/acra-translator/acra-translator.go [hunk 2]
```diff
@@ -80,6 +83,12 @@ func main() {
...
+	tlsIdentifierExtractorType := flag.String("tls_identifier_extractor_type", network.IdentifierExtractorTypeDistinguishedName, fmt.Sprintf("Decide which field of TLS certificate to use as ClientID (%s)", strings.Join(network.IdentifierExtractorTypesList, "|")))
+	useClientIDFromConnection := flag.Bool("acratranslator_client_id_from_connection_enable", false, "Use clientID from TLS certificates or secure session handshake instead directly passed values in grpc methods")
```
Comment (vixentael) [inline cmd/acra-translator/acra-translator.go:88]:
```suggestion
	useClientIDFromConnection := flag.Bool("acratranslator_client_id_from_connection_enable", false, "Use clientID from TLS certificates or secure session handshake instead directly passed values in gRPC methods")
```

File: cmd/acra-translator/acra-translator.go [hunk 3]
```diff
@@ -80,6 +83,12 @@ func main() {
...
+	noEncryptionTransport := flag.Bool("acraconnector_transport_encryption_disable", false, "Use raw transport (tcp/unix socket) between AcraTranslator and client. It turns off reading trace from client's side which usually sent by AcraConnector")
```
Comment (vixentael) [inline cmd/acra-translator/acra-translator.go:89]:
```suggestion
	noEncryptionTransport := flag.Bool("acraconnector_transport_encryption_disable", false, "Use raw transport (tcp/unix socket) between AcraTranslator and client app. It turns off reading trace from client app's side which usually sent by AcraConnector")
```

File: cmd/acra-translator/acra-translator.go [hunk 4]
```diff
@@ -131,13 +142,61 @@ func main() {
...
+			os.Exit(1)
+		}
+		clientIDExtractor, err = network.NewTLSClientIDExtractor(identifierExtractor, idConverter)
+		if err != nil {
+			log.WithError(err).Errorln("Can't initialize clientID extractor")
+			os.Exit(1)
+		}
+		grpcClientIDExtractor, err = network.NewTLSGRPCClientIDExtractor(clientIDExtractor)
+		if err != nil {
+			log.WithError(err).Errorln("Can't initialize grpc client id extractor")
```
Comment (vixentael) [inline cmd/acra-translator/acra-translator.go:173]:
```suggestion
			log.WithError(err).Errorln("Can't initialize gRPC client id extractor")
```

File: cmd/acra-translator/grpc_api/factory.go [hunk 5]
```diff
@@ -32,12 +32,23 @@ type GRPCServerFactory struct{}
...
 		logrus.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorTranslatorCantHandleGRPCConnection).
 			Errorln("Can't create grpc service")
 		return nil, err
 	}
+	if data.UseConnectionClientID {
+		logrus.Infoln("Wrap grpc service to use clientID from connection")
+		service, err = NewTLSDecryptServiceWrapper(data.ConnectionClientIDExtractor, service)
+		if err != nil {
+			logrus.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorTranslatorCantHandleGRPCConnection).
+				Errorln("Can't wrap grpc service with TLS wrapper")
```
Comment (vixentael) [inline cmd/acra-translator/grpc_api/factory.go:48]:
```suggestion
				Errorln("Can't wrap gRPC service with TLS wrapper")
```

**Review by vixentael — COMMENTED** [review]
Sorry for me being boring :)

File: network/tls_authentication.go [hunk 6]
```diff
@@ -102,6 +103,36 @@ func (c hexIdentifierConverter) Convert(identifier []byte) ([]byte, error) {
 	return out, nil
 }
 
+// TLSClientIDExtractor complex component for extracting clientID from certificates using certificate identifier extractor and convertor
+type TLSClientIDExtractor interface {
+	ExtractClientID(certificate *x509.Certificate) ([]byte, error)
+}
```
Comment (ilammy) [inline network/tls_authentication.go]: I'm happy to see these two things merged. Though, this capitalization is kinda confusing to read: TLS—Client—IDE—xtractor. I constantly have to correct myself that this has nothing to do with IDEs.

I do not have any alternatives to suggest, unfortunately.

File: network/tls_wrapper.go [hunk 7]
```diff
@@ -197,3 +199,28 @@ func (wrapper *TLSConnectionWrapper) WrapServer(ctx context.Context, conn net.Co
...
 	return newSafeCloseConnection(tlsConn), clientID, nil
 }
+
+type tlsGRPCClientIDExtractor struct {
+	extractor TLSClientIDExtractor
+}
+// NewTLSGRPCClientIDExtractor return GRPCConnectionClientIDExtractor implementation which extracts info from peer's tls certificate
+func NewTLSGRPCClientIDExtractor(extractor TLSClientIDExtractor)(*tlsGRPCClientIDExtractor, error){
+	return &tlsGRPCClientIDExtractor{extractor: extractor}, nil
+}
```
Comment (ilammy) [inline network/tls_wrapper.go]: This AaAO (Acronym and Abbreviation Overload) really reads like `New-ЦК-КПСС-lient-IDE-трактор` to me.

**Review by ilammy — COMMENTED** [review]
(Haven't read through the tests yet.)

**Review by shadinua — APPROVED** [review]

File: network/tls_wrapper.go (hunk 7 above)
Comment (Lagovas) [reply network/tls_wrapper.go]: another way that I know to save readability is split to several subpackages... create subpackage `grpc` in `network` and there `TLSClientIDExtractor` or create one more subpackages `tls` + `secure_session` where will be `ClientIDExtractor`....

File: cmd/acra-translator/grpc_api/decryptor_test.go [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+
+	go server.Serve(listener)
+	return &testgRPCServer{listener: listener, server: server, unixPath: unixPath.Name()}
+}
+
+// newServerTLSgRPCOpts returns server options to use TLS as transport using keys from tests/ssl/[ca|acra-server]
+func newServerTLSgRPCOpts(t *testing.T) []grpc.ServerOption {
+	verifier := network.NewCertVerifierAll()
+	workingDirectory := tests.GetSourceRootDirectory(t)
+	serverConfig, err := network.NewTLSConfig("localhost", filepath.Join(workingDirectory, "tests/ssl/ca/ca.crt"), filepath.Join(workingDirectory, "tests/ssl/acra-server/acra-server.key"), filepath.Join(workingDirectory, "tests/ssl/acra-server/acra-server.crt"), 4, verifier)
```
Comment (iamnotacake) [inline cmd/acra-translator/grpc_api/decryptor_test.go:71]: Such a long line, don't you want to split it, one arg per line, for the sake of readability?

File: cmd/acra-translator/server/server.go [hunk 9]
```diff
@@ -211,7 +215,7 @@ func (server *ReaderServer) Start(parentContext context.Context) {
 	server.detectPoisonRecords(poisonCallbacks)
 	errCh := make(chan error)
 
-	decryptorData := &common.TranslatorData{Keystorage: server.keystorage, PoisonRecordCallbacks: poisonCallbacks, CheckPoisonRecords: server.config.DetectPoisonRecords()}
+	decryptorData := &common.TranslatorData{Keystorage: server.keystorage, PoisonRecordCallbacks: poisonCallbacks, CheckPoisonRecords: server.config.DetectPoisonRecords(), UseConnectionClientID: server.config.GetUseClientIDFromConnection(), ConnectionClientIDExtractor: server.config.GetGRPCClientIDExtractor()}
```
Comment (iamnotacake) [inline cmd/acra-translator/server/server.go:218]:
```suggestion
	decryptorData := &common.TranslatorData{
		Keystorage: server.keystorage,
		PoisonRecordCallbacks: poisonCallbacks,
		CheckPoisonRecords: server.config.DetectPoisonRecords(),
		UseConnectionClientID: server.config.GetUseClientIDFromConnection(),
		ConnectionClientIDExtractor: server.config.GetGRPCClientIDExtractor(),
	}
```

File: cmd/acra-translator/server/server.go [hunk 10]
```diff
@@ -348,7 +352,7 @@ func (server *ReaderServer) StartFromFileDescriptor(parentContext context.Contex
...
```
Comment (iamnotacake) [inline cmd/acra-translator/server/server.go:355]:
```suggestion
	decryptorData := &common.TranslatorData{
		Keystorage: server.keystorage,
		PoisonRecordCallbacks: poisonCallbacks,
		CheckPoisonRecords: server.config.DetectPoisonRecords(),
		UseConnectionClientID: server.config.GetUseClientIDFromConnection(),
		ConnectionClientIDExtractor: server.config.GetGRPCClientIDExtractor(),
	}
```

**Review by iamnotacake — APPROVED** [review]

**Comment by Lagovas** [comment]
@vixentael , I applied your suggestions: https://github.com/cossacklabs/acra/pull/444/commits/1770439e35672492ba8561b8ca905ea0cc87fe52
@iamnotacake, same: https://github.com/cossacklabs/acra/pull/444/commits/9a07c0337cc87b8a4bbb5570a0eb4c06a0cb4655

**Review by vixentael — APPROVED** [review]
Thank you, LGTM! 👏

**Pull request merged**